tests/mocha/test_helpers.js

```diff
@@ -81,6 +84,86 @@ function createEventsFireStub() {
   return stub;
 }
 
+/**
+ * Shared setup method that sets up fake timer for clock so that pending
+ * setTimeout calls can be cleared in test teardown along with other common
+ * stubs. Should be called in setup of outermost suite using
+ * sharedTestSetup.call(this).
+ * @param {Object<string, boolean>} options Options to enable/disable setup
+ *    of certain stubs.
+ */
+function sharedTestSetup(options = {}) {
+  this.sharedSetupCalled_ = true;
+  // Sandbox created for greater control when certain stubs are cleared.
+  this.sharedSetupSandbox_ = sinon.createSandbox();
+  this.clock = this.sharedSetupSandbox_.useFakeTimers();
+  if (options['fireEventsNow'] === undefined || options['fireEventsNow']) {
+    // Stubs event firing unless passed option "fireEventsNow: false"
+    this.eventsFireStub = createEventsFireStubFireImmediately_(this.clock);
+  }
+}
+
+/**
+ * Shared cleanup method that clears up pending setTimeout calls, disposes of
+ * workspace, and resets global variables. Should be called in setup of
+ * outermost suite using sharedTestTeardown.call(this).
+ */
+function sharedTestTeardown() {
+  if (!this.sharedSetupCalled_) {
+    console.error('"' + this.currentTest.fullTitle() +
+        '" did not call sharedTestSetup');
+  }
+
+  try {
+    if (this.workspace) {
+      workspaceTeardown.call(this, this.workspace);
+      this.workspace = null;
+    } else {
+      this.clock.runAll();  // Run all queued setTimeout calls.
+    }
+  } catch (e) {
+    console.error(this.currentTest.fullTitle() + '\n', e);
+  } finally {
+    // Clear Blockly.Event state.
+    Blockly.Events.setGroup(false);
+    Blockly.Events.disabled_ = 0;
+    if (Blockly.Events.FIRE_QUEUE_.length) {
+      // If this happens, it may mean that some previous test is missing cleanup
+      // (i.e. a previous test added an event to the queue on a timeout that
+      // did not use a stubbed clock).
+      Blockly.Events.FIRE_QUEUE_.length = 0;
+      console.warn(this.currentTest.fullTitle() +
+          '" needed cleanup of Blockly.Events.FIRE_QUEUE_. This may indicate ' +
+          'leakage from an earlier test');
+    }
+
+    // Restore all stubbed methods.
+    this.sharedSetupSandbox_.restore();
+    sinon.restore();
```

Comment:
Why also call `sinon.restore`? It looks like everything done in this shared setup used the sandbox.

Reply:
The only stub on `sharedSetupSandbox_` is `this.clock`. All the stubs defined in the tests define themselves on `sinon` directly, which is why I need to also do `sinon.restore`. I have the clock on a separate sandbox so that if anywhere else `sinon.restore` is called, it doesn't affect the stubbed clock.

Reply:
Ah, I see--the events stub is not on the sandbox.

Reply:
That, and also other stubs and spies in individual tests.

Reply:
Right--the stubs and spies in individual tests are an argument for calling restore in the same place where they set up the stub or spy, though. My concern is around the asymmetry, where we can sometimes expect to be cleaning up state in this teardown that the matching setup function did not create.